### PR TITLE
Queue deployment

### PR DESCRIPTION
**Pull request by tiopramayudi**
**What this PR does / why we need it**:

Add queue mechanism when deploying inferenceservice & batch job. This is to solve issue when merlin server is restarted by external factor and the ongoing deployment(inferenceservice or batch) will be failed and status is not updated
**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

**Checklist**

- [x] Added unit test, integration, and/or e2e tests
- [ ] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -54,3 +54,4 @@ MONITORING_DASHBOARD_JOB_BASE_URL=http://monitoring.dev/batch-job-dashboard
 MLP_API_ENCRYPTION_KEY=password
 
 # MLFLOW_TRACKING_URL=
+NUM_OF_WORKERS=1
```
Comment (pradithya) [inline .env.sample:57]: What's the ideal number of worker?

File: api/cluster/controller.go [hunk 2]
```diff
@@ -156,6 +158,13 @@ func (k *controller) Deploy(modelService *models.Service) (*models.Service, erro
 
 	s, err = k.waitInferenceServiceReady(s)
 	if err != nil {
+		// remove newly created inferenceservice when got error
```
Comment (pradithya) [inline api/cluster/controller.go:161]: This is unrelated to queue implementation right? it's for the fix when deployment is considered fail but we didn;t garbage collect the resource.

File: api/cmd/ui_handler.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
+package main
```
Comment (pradithya) [inline api/cmd/ui_handler.go:1]: Does it make sense to have separate folder for cmd.go, setup.go, and ui_handler.go ? I.e. under cmd/api/

File: api/queue/dispatcher.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+	"github.com/gojek/merlin/log"
+	"github.com/jinzhu/gorm"
+)
+
+type Producer interface {
+	EnqueueJob(job *Job) error
+}
+
+type Consumer interface {
```
Comment (pradithya) [inline api/queue/dispatcher.go:18]: Can you add description to these interfaces : Producer and Consumer

File: api/queue/dispatcher.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+type JobFn func(*Job) error
+
+type Config struct {
+	NumWorkers int
+	Db         *gorm.DB
+}
+
+type Status string
+
+type Argument map[string]interface{}
```
Comment (pradithya) [inline api/queue/dispatcher.go:41]: Maybe rename it to `Arguments` since it contains more than one Argument

File: api/queue/dispatcher.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+type Job struct {
+	ID        int64     `json:"id" gorm:"primary_key;"`
+	Arguments Argument  `json:"arguments"`
+	Completed bool      `json:"completed"`
+	Name      string    `json:"name"`
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
+}
+
+func (a Argument) Value() (driver.Value, error) {
```
Comment (pradithya) [inline api/queue/dispatcher.go:52]: I think `Arguments` definition should be separated into its own file.

File: api/queue/dispatcher.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		return errors.New("type assertion to []byte failed")
+	}
+
+	return json.Unmarshal(b, &a)
+}
+
+func NewDispatcher(cfg Config) *Dispatcher {
+	workers := make([]*worker, 0, cfg.NumWorkers)
+	jobChan := make(chan *Job)
+	jobFuncMap := make(map[string]JobFn)
```
Comment (pradithya) [inline api/queue/dispatcher.go:68]: I think jobFuncMap should use the thread-safe version of map instead of stdlib map to avoid concurrency issue

File: api/queue/dispatcher.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	}
+}
+
+func (d *Dispatcher) EnqueueJob(job *Job) error {
+	var savedJob Job
+	if err := d.db.Save(job).Scan(&savedJob).Error; err != nil {
+		log.Errorf("Failed to save job %d with error: %v", job.ID, err)
+		return err
+	}
+	go func() {
```
Comment (pradithya) [inline api/queue/dispatcher.go:87]: What is this code for?

File: api/queue/worker.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+				go w.processJob(job)
+			}
+		}
+	}()
+}
+
+func (w *worker) processJob(job *Job) {
+	var refreshedJob Job
+	tx := w.db.Begin()
+	err := tx.Raw(findJobQuery, job.ID, false).Scan(&refreshedJob).Error
```
Comment (pradithya) [inline api/queue/worker.go:47]: Is it possible to not use `SKIP_LOCKED` in the query? The reason is that `worker` become unable to identify whether the row already has a lock or really doesn't exists.

File: api/service/model_endpoint_service.go [hunk 10]
```diff
@@ -35,6 +35,9 @@ import (
 )
 
 const (
+	RealtimeDeploymentJob = "realtime"
```
Comment (pradithya) [inline api/service/model_endpoint_service.go:38]: I think `model service deployment` is a better name

File: api/service/prediction_job_service.go [hunk 11]
```diff
@@ -42,6 +43,8 @@ type PredictionJobService interface {
 	ListContainers(env *models.Environment, model *models.Model, version *models.Version, predictionJob *models.PredictionJob) ([]*models.Container, error)
 	// StopPredictionJob deletes the spark application resource and cleans up the resource
 	StopPredictionJob(env *models.Environment, model *models.Model, version *models.Version, id models.ID) (*models.PredictionJob, error)
+	// ExecuteDeployment creates prediction job based on queue job that produce by CreatePredictionJob method
```
Comment (pradithya) [inline api/service/prediction_job_service.go:46]: I think `ExecuteDeployment` shouldn't be part of `PredictionJobService`.  
`PredictionJobService` should just submit the request to queue, and then queue+worker handle the rest.

File: api/service/version_endpoint_service.go [hunk 12]
```diff
@@ -60,24 +75,121 @@ type endpointService struct {
...
+
+type EndpointJobArgs struct {
+	Endpoint *models.VersionEndpoint
+	Model    *models.Model
+	Version  *models.Version
+	Project  mlp.Project
+}
+
+// ExecuteDeployment will create inferenceservice based on job that already queued
+func (k *endpointService) ExecuteDeployment(job *queue.Job) error {
```
Comment (pradithya) [inline api/service/version_endpoint_service.go:102]: Similar to comment https://github.com/gojek/merlin/pull/107/files#r601994686 . I think `ExecuteDeployment` shouldn't be part of the service. It's better to split it as separate class and closely related to `queue` package.

File: db-migrations/1_schema_creation.up.sql [hunk 13]
```diff
@@ -16,7 +16,7 @@ CREATE TYPE endpoint_status as ENUM ('pending', 'running', 'serving', 'failed',
 
 CREATE TABLE IF NOT EXISTS projects
 (
-    id                  serial PRIMARY KEY,
+    id                  bigserial PRIMARY KEY,
```
Comment (pradithya) [inline db-migrations/1_schema_creation.up.sql:19]: Will it become an issue?

File: api/cluster/controller.go (hunk 2 above)
Comment (tiopramayudi) [reply api/cluster/controller.go:161]: yes, it is not related to queue implementation. I changed this also because it is related to model deployment

File: api/cmd/ui_handler.go (hunk 3 above)
Comment (tiopramayudi) [reply api/cmd/ui_handler.go:1]: yes make sense

File: db-migrations/1_schema_creation.up.sql (hunk 13 above)
Comment (tiopramayudi) [reply db-migrations/1_schema_creation.up.sql:19]: that is not supposed to be here? will revert this

File: api/queue/dispatcher.go (hunk 8 above)
Comment (tiopramayudi) [reply api/queue/dispatcher.go:87]: so for enqueuing job, queue will insert the job record to DB if it success it will send the job into channel, since the worker will consume job from values from channel

File: api/queue/worker.go (hunk 9 above)
Comment (tiopramayudi) [reply api/queue/worker.go:47]: there will be no case the row will be empty because the record is not there in DB because all job already persist to DB before. I think case that I havent covered is when query return err because of issue with DB. I think for this case we can requeue the job after delay for several seconds

File: api/service/version_endpoint_service.go (hunk 12 above)
Comment (tiopramayudi) [reply api/service/version_endpoint_service.go:102]: package queue will handle registration of job and consume the job, but the logic to consume the job itself is handled in relevant package. IMO version_endpoint_service or prediction_job_service still relevant to deploy the job, or do we need another file for deployment only let say deployment_service.go?

File: api/service/model_endpoint_service.go (hunk 10 above)
Comment (tiopramayudi) [reply api/service/model_endpoint_service.go:38]: ok I'll change this

File: api/queue/dispatcher.go (hunk 6 above)
Comment (tiopramayudi) [reply api/queue/dispatcher.go:52]: will change this

File: api/queue/worker.go (hunk 9 above)
Comment (pradithya) [reply api/queue/worker.go:47]: Then what is this code for?
```
	if err == gorm.ErrRecordNotFound {
		tx.Rollback()
		log.Warnf("Job records not found with id: %d", job.ID)
		return
	}
```

File: api/service/version_endpoint_service.go (hunk 12 above)
Comment (pradithya) [reply api/service/version_endpoint_service.go:102]: Yeah, better to separate these into its own package. probably under queue/work. We can name it deployment.go and batch_prediction.go

File: api/queue/worker.go (hunk 9 above)
Comment (tiopramayudi) [reply api/queue/worker.go:47]: to return early, this indicate that job already success or locked in other process. The logging message should be improve, I'll change this

File: api/service/version_endpoint_service.go [hunk 14]
```diff
@@ -60,26 +73,123 @@ type endpointService struct {
...
 }
 
+type EndpointJobArgs struct {
+	Endpoint *models.VersionEndpoint
+	Model    *models.Model
+	Version  *models.Version
+	Project  mlp.Project
+}
+
+// ExecuteDeployment will create inferenceservice based on job that already queued
```
Comment (pradithya) [inline api/service/version_endpoint_service.go:99]: Dead code

File: api/work/batch_deployment.go [hunk 15]
```diff
@@ -0,0 +1,85 @@
+package work
```
Comment (pradithya) [inline api/work/batch_deployment.go:1]: Can it be part of /queue/work package ?

File: api/models/batch_job.go [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package models
+
+import "github.com/gojek/merlin/mlp"
+
+type BatchJob struct {
```
Comment (pradithya) [inline api/models/batch_job.go:5]: I think we shouldn't put `BatchJob` and `EndpointJob` under `models` package. It's better to put it closer to the `work` package

File: api/models/endpoint_job.go [hunk 17]
```diff
@@ -0,0 +1,12 @@
+package models
+
+import (
+	"github.com/gojek/merlin/mlp"
+)
+
+type EndpointJob struct {
```
Comment (pradithya) [inline api/models/endpoint_job.go:7]: https://github.com/gojek/merlin/pull/107/files#r603758002

**Review by pradithya — APPROVED** [review]
LGTM
